platforms/defaults_windows_test.go

```diff
@@ -241,12 +247,17 @@ func TestMatchComparerLess(t *testing.T) {
 		{
 			Architecture: "amd64",
 			OS:           "windows",
-			OSVersion:    "10.0.17763.2",
+			OSVersion:    fmt.Sprintf("10.0.17763.%d", ubr), // version with same UBR as host is best match
```

Comment:
I'm not sure I agree it should be "exact match then highest to lowest". Its always safe to run the higher patch set right? Why wouldnt we want the latest patches for the matching major os version?

Reply:
I agree see the part after the HR in https://github.com/containerd/containerd/issues/6508#issuecomment-1029632672.

Reply:
It is always safe to run the higher patch (except for 2016), but not always the best option in respect of performance.

Latest patch is the current behavior. Basically there is no point currently in building a multi arch image with same OS version and different UBRs, since highest UBR always wins. However, if as a developer I made the decision to create a multi-arch window image with same OS version and different UBRs, I would expected the best host match be used. 

Reason for this is that pulling Windows docker images has a big impact on pod startup time. 

Typically k8s nodes from major cloud providers come with a Windows server core and nano image to speed up Windows pod startup. However, with different k8s deployments the UBR of the nodes can slightly differ, which may cause difference in image pull time in minute range for missing layers if there is only one image available.

Building a multi arch windows image with all supported UBRs would allow to get consistent pull performance on all k8s deployments, and none of the OS layers needs to be pulled. So this an example from the "wild" where using different UBRs in a multi-arch windows can make sense (#6508). In our use case we use argo workflows, with many windows pod run on demand basis. Pod startup time makes a difference in the run time of the workflow.

Deploying individual images with different tags is a workaround, but from a deployment perspective much more complex.

Reply:
I think this is a different problem: your use-case is to prefer an acceptable and already-partially-present image over pulling a different (and potentially more-preferable image), when evaluating a manifest list.

_This_ approach to achieve that conflates "already partially present" with "same UBR as the host", which is presumably specific to the provider or providers doing this layer pre-caching, only applies to the OS-specific base layer, and would lead to a worse outcome for everyone who doesn't have such pre-pulls done for them (or has such pre-pulls, but then updates the node to a newer patch release). And the change fails to achieve its purpose if the pre-cached image doesn't happen to match the UBR of the host, e.g., pointing again at the Feb 2020 security update where if the providers were going to pre-cache a base image, it should be the newer container base with the security patches, not the one that matched the host OS (and didn't run 32-bit binaries correctly).

Similar to HyperV-Isolation-compatible and LCOW-compatible version matchers, my feeling is that this proposed change should be an external platform matcher used when the caller knows it is the desired behaviour (due to some host-specific behaviour in this case), not the default matcher. See the discussion (particularly at the end) on #6491 about caller-provided matchers.

I'd suggest a feature request or something for the actual use-case here, as the proposed change works against the existing discussion and use-cases. "Prefer already-pulled image layers when choosing from a manifest list" is an interesting feature, I'd be interested to see how that can be achieved in a general sense, rather than specifically for Windows base layers. I personally wouldn't want it on-by-default, perhaps it'd be another value in the k8s `imagePullPolicy`, and/or an annotation that comes through via CRI's `PullImage` API.

For changes to the defult version-matching logic specifically, it might be worth carrying this discussion back to #6508, since it touched upon the same topics, but since that ticket is more about OS version matching, the UBR-matching notes there were more of a "I've been thinking about this, and here's where I happened to be when the topic came up". Otherwise, discussion on #6693 might be easier to follow than a thread on a PR review.